thedmstoolkit/character_generation.py

```diff
@@ -0,0 +1,538 @@
+        Args:
+            background_list (str): background
+            GeneratorKey (str): key required to specify which randomizer you want
+
+        Raises:
+            RuntimeError: If the key provided is the wrong type
+            RuntimeError: If there is an unauthorized item in the passed list
+
+        Returns:
+            String: Background
+        """
+        if GeneratorKey not in Character_Generator.Generators:
+            raise RuntimeError("Generator does not exist")
+        for x in background_list:
+            if x not in Character_Generator.BACKGROUND_LIST:
+                raise RuntimeError("Invalid List")
+        return background_list[
+            Character_Generator.Generators[GeneratorKey](0, len(background_list) - 1)
+        ]
+
+    @staticmethod
+    def Generate(
+        generations_list: Optional[list[str]] = None,
+        stat_generator_key: Optional[str] = None,
+        race_key="All",
+        class_key="All",
+        alignment_key="All",
+        GeneratorKey="random",
+        stat_list=Optional[list[int]],
+    ) -> dict[str, Union[list[int], str]]:
+        """Given generator parameters, return a dictionary of character characteristics
+
+        Args:
+            generations_list (str, optional): A list of the various generations to do. Defaults to "All".
+            stat_generator_key (Optional[str], optional): A generator name from limited_generators(). Defaults to None.
+            race_key (str, optional): A race category, from the RACE_DICT. Defaults to "All".
+            class_key (str, optional): A class category from the CLASS_DICT. Defaults to "All".
+            alignment_key (str, optional): A alignment category from the ALIGNMENT_DICT. Defaults to "All".
+            stat_list (list[int], optional): A list of 6 numbers between 1 and 20 to use as stats. Defaults to None.
+
+        Raises:
+            ValueError: _description_
+            RuntimeError: Not all numbers are in the passed stat array are valid
+
+        Returns:
+            dict[str, list[int] | str]: Generated Characteristics
+        """
+
+        if type(stat_list) is list:
+            if len(stat_list) != 6:
+                raise ValueError("Stat list length must be 6")
+
+            for i in stat_list:
+                if type(i) != int:
+                    raise RuntimeError("Not all numbers in Stat array are integers")
+                if not 0 < i <= 18:
+                    raise RuntimeError(
+                        "There is an invalid number in the array, please use 1-18"
+                    )
+        if stat_generator_key is None:
+            stat_generator_keys: list[str] = list(
+                Character_Generator.get_all_random_generators()
+            )
+            stat_generator_key = stat_generator_keys[
+                random.randint(0, len(stat_generator_keys) - 1)
+            ]
+        # Set the stat generator if selected
+
+        if generations_list is None:
```

Comment:
do we check if `generations_list` contains invalid elements?

Reply:
I deemed it unnecessary because it just ignores any non accepted strings, I can add it though